src/smt/smt_quantifier.cpp

```diff
@@ -108,38 +109,39 @@ namespace smt {
            \brief Ensures that all relevant proof steps to explain why the enode is equal to the root of its
            equivalence class are in the log and up-to-date.
         */
-        void log_justification_to_root(std::ostream & log, enode *en) {
+        void log_justification_to_root(std::ostream & log, enode *en, obj_hashtable<enode> &already_visited) {
             enode *root = en->get_root();
             for (enode *it = en; it != root; it = it->get_trans_justification().m_target) {
-                if (it->m_proof_logged_status == smt::logged_status::NOT_LOGGED) {
-                    it->m_proof_logged_status = smt::logged_status::BEING_LOGGED;
-                    log_single_justification(log, it);
-                    it->m_proof_logged_status = smt::logged_status::LOGGED;
-                } else if (it->m_proof_logged_status != smt::logged_status::BEING_LOGGED && it->get_trans_justification().m_justification.get_kind() == smt::eq_justification::kind::CONGRUENCE) {
-
-                    // When the justification of an argument changes m_proof_logged_status is not reset => We need to check if the proofs of all arguments are logged.
-                    it->m_proof_logged_status = smt::logged_status::BEING_LOGGED;
+                if (already_visited.find(it) == already_visited.end()) already_visited.insert(it);
+                else break;
+
+                if (!it->m_proof_is_logged) {
+                    log_single_justification(log, it, already_visited);
+                    it->m_proof_is_logged = true;
+                } else if (it->get_trans_justification().m_justification.get_kind() == smt::eq_justification::kind::CONGRUENCE) {
```

Comment:
is this correct if commutativity was used? There is code below in a separate case that singles out commutativity as nyi

Reply:
Yes, that would be a problem. Z3's E-graph only supports commutativity for binary functions, right? In that case it would be easy to add full support in another pull request. How would one cause such a justification to be generated? Is there a way to declare functions commutative in the SMT-file or is this just supported for built in functions?

Thanks,
Nils